### PR TITLE
[GOLD] Tool calling support for GOLDTrainer

### PR DESCRIPTION
**Pull request by Strongich**
# What does this PR do?

Adds **multi-turn tool-calling support** to the `GOLDTrainer`, for **same-family** (shared-tokenizer) distillation. 

This addition was motivated by these 2 blogs about frontier models' training: [first](https://huggingface.co/blog/sergiopaniego/distillation-2026) and [second](https://maximelabonne.substack.com/p/the-state-of-the-open-frontier). 

This PR wires tool calling through both paths: off and on policy, for text LLMs and same-family VLMs, and ports GRPO's `environment_factory` so environment methods become callable tools.

## What's included

- **Config**: `max_tool_calling_iterations` (cap on tool-loop turns per rollout); `GOLDConfig` now overrides `remove_unused_columns` to default `False` (like `GRPOConfig`) so custom dataset columns reach `environment.reset(**kwargs)`.
- **Off-policy masking** (`DataCollatorForChatML` / `DataCollatorForVisionLanguageChatML`, and the pretokenized dataset-prep path): the prompt is masked, assistant turns — including the assistant's tool-*call* tokens — are supervised, and tool-*result* tokens are masked to `-100`. Mirrors `sft_trainer.py` semantics.
- **On-policy tool loop**: `_tool_call_loop` / `_get_tool_suffix_ids` / `_generate_single_turn` ported near-verbatim from `grpo_trainer.py`. A `tool_mask` tracked during generation excludes tool-result tokens from the JSD/Liger loss.
- **`environment_factory`**: ported from GRPO (single-environment form) — pooled per-rollout instances, `reset(**example)` per rollout, observations injected back into the conversation. `get_reward` is excluded from tool extraction, matching GRPO's reserved env-owned-reward hook (#6238).
- **Same-family only, enforced at init**: `tools` + `use_uld_loss=True`, `tools` + `seq_kd=True`, and `tools` + `packing=True` raise `ValueError`; VLM tool rollouts require `use_vllm=True`; multimodal tool *responses* (tools returning images) raise `NotImplementedError`.

## Example scripts

Two runnable examples, both **pure on-policy** (`lmbda=1.0`): the student generates and *actually executes* the tools, and the teacher scores the student's own rollouts.

- **`gold_tool_calling.py`** (text): Qwen3-8B → Qwen3-1.7B on a browser-agent dataset ([`DataCreatorAI/tool-calling-browser-agent-tasks`](https://huggingface.co/datasets/DataCreatorAI/tool-calling-browser-agent-tasks)), filtered to conversations using six executable browser-form tool stubs passed via `tools=`. Demonstrates the plain-`tools` path: shared tokenizer, Qwen3's native tool-calling template, thinking disabled via `chat_template_kwargs`.
- **`gold_tool_calling_vlm.py`** (VLM): Qwen3-VL-8B-Instruct → Qwen3-VL-2B-Instruct on Search-VL VQA tasks ([`OpenSearch-VL/Search-VL-SFT-36K`](https://huggingface.co/datasets/OpenSearch-VL/Search-VL-SFT-36K)), with live tools: DuckDuckGo search (`smolagents`) plus a docling-based `layout_parsing` environment — this one exercises `environment_factory` (pooled instances, per-example `reset` resolving image references). Tool outputs are capped and the tool loop is bounded by `max_tool_calling_iterations=10`. 

## Notable details

- **Prefix-preserving template**: multi-turn tool masking locates message boundaries by incremental prefix rendering, which requires a prefix-preserving chat template. When the tokenizer's own template isn't (e.g. Qwen3 drops historical `<think>` blocks), all tool-flow renders — off-policy prep, both collators, and the on-policy `_get_tool_suffix_ids` — use a single training-safe template computed once at init.
- **VLM + vLLM image-token expansion** ([#6294](https://github.com/huggingface/trl/issues/6294)): the vLLM generation path intentionally matches GRPO's as-is (consistency over correctness), so it inherits this open issue rather than carrying a GOLD-only workaround. The bug is model-dependent and does not affect Qwen-VL, which both examples use; whatever fix lands for GRPO will apply here unchanged.

## Testing

- New off-policy masking, on-policy tool-loop smoke, `environment_factory`, and validation-guard tests in `tests/experimental/test_gold_trainer.py`.
- Both example scripts trained end-to-end on a single 80GB A100 (transformers 5.x + vLLM colocate + Liger fused JSD).

## Before submitting

- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request), Pull Request section?
- [ ] Was this discussed/approved via a GitHub issue?
- [X] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## AI writing disclosure

- [x] AI-assisted: some parts were suggested or improved by AI, but the PR was written and reviewed by a human.

---

> [!NOTE]
> **Medium Risk**
> Touches on-policy generation, loss masking, and dataset collation in the experimental GOLD training path; constraints are validated at init but multi-turn tool loops add complexity and new failure modes (length limits, template drift).
> 
> **Overview**
> Adds **multi-turn tool-calling distillation** to `GOLDTrainer` for same-family student/teacher setups, aligned with GRPO’s agent API (`tools`, optional `environment_factory`).
> 
> **On-policy:** the student generates tool calls, executes them (sync/async), and the teacher distills on the trajectory; `tool_mask` keeps tool-*result* tokens out of the loss while supervising assistant/tool-call tokens. **Off-policy** (`lmbda < 1`) requires dataset `tools` + full trajectories; prep and collators emit/apply `tool_mask`, with prefix-preserving chat templates when needed (e.g. Qwen3).
> 
> **Config/behavior:** `max_tool_calling_iterations`, `remove_unused_columns=False` by default for `environment.reset(**kwargs)`; incompatible combos (`use_uld_loss`, `seq_kd`, `packing`, multimodal tool responses) fail at init; VLM tool rollouts require `use_vllm=True`.
> 
> Documentation (`gold_trainer.md`), runnable `gold_tool_calling.py` / `gold_tool_calling_vlm.py` examples, and broad tests in `test_gold_trainer.py` cover validation, masking, smoke training, and `environment_factory`.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 797a98d416b0c9abc5f510db8806dde6a6f4b9d4. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: trl/experimental/utils.py [hunk 1]
```diff
@@ -656,6 +792,31 @@ def _raw_text_from_messages(messages_or_str: Any) -> str:
...
+                prompt_messages + completion_messages[: k + 1],
+                tools=self.tools,
+                tokenize=False,
+                chat_template=self.chat_template,
+            )
+            continuation = rendered[len(rendered_prompt) :]
+            num_tokens = len(self.processor(text=[continuation], add_special_tokens=False)["input_ids"][0])
+            if completion_messages[k].get("role") == "tool":
+                ranges.append((prev_tokens, num_tokens))
+            prev_tokens = num_tokens
```
[the remark this answers is not in the document]
Comment (Strongich) [reply trl/experimental/utils.py:827]: same mechanic from GRPO `_get_tool_suffix_ids`, FP

File: trl/experimental/gold/gold_trainer.py [hunk 2]
```diff
@@ -1637,6 +1821,504 @@ def _generate_on_policy_for_slices(
...
+                else:
+                    sync_tool_dict[tool.__name__] = tool
+            self._sync_tool_dicts.append(sync_tool_dict)
+            self._async_tool_dicts.append(async_tool_dict)
+
+        if self.environments:
+            # `DataCollatorForChatML(forward_prompt_messages=True)` forwards the raw example dicts as `reset_kwargs`,
+            # so `reset` receives the same per-rollout kwargs as the VLM path and GRPO.
+            for prompt, environment, reset_kwargs in zip(prompts, self.environments, reset_kwargs_list, strict=True):
+                observation = environment.reset(**reset_kwargs)
```
[the remark this answers is not in the document]
Comment (Strongich) [reply trl/experimental/gold/gold_trainer.py:1877]: both `out["prompt"]` and `out["reset_kwargs"]` are emitted by `DataCollatorForChatML(forward_prompt_messages=True)` together with length = batch size. FP

**Comment by Strongich** [comment]
Hi @kashif! I've added multi-turn tool-calling support to this trainer -- would appreciate a review when you have time.

Scope: **same-family** LMs and VLMs only. For VLMs, tool rollouts additionally require `use_vllm=True`.

Two examples added, both pure on-policy (lmbda=1.0) with use_vllm=True:

1. Text: Qwen3-8B -> Qwen3-1.7B (trl/experimental/gold/gold_tool_calling.py)
image

2. VLM: Qwen3-VL-8B -> Qwen3-VL-2B (trl/experimental/gold/gold_tool_calling_vlm.py)
image

A few caveats on the examples:

1. The VLM run drifted: as I found while experimenting afterwards, the off-the-shelf teacher itself is not stable with the system prompt and the stand-in tools we use (it often answers from memory instead of calling them), so on-policy distillation pulled the student from ~2 tool calls per rollout initially down to ~0. Producing a high-level example would require fine-tuning the teacher to use these tools first, and then either generating traces to allow $\lambda < 1$ or continuing full on-policy distillation with the fine-tuned teacher -- I don't have enough time & compute for either. Nevertheless, tool distillation is a niche training approach, and the general idea of this PR is to give users who already know what they are doing the tool they can use for it with zero (or almost zero) need to write additional trainer code.
2. Text-only distillation is more stable in terms of training, but the tools used are mocks (the chosen dataset's tools would require real form-submission backends).

File: trl/experimental/gold/gold_tool_calling_vlm.py [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+
+        Returns:
+            The extracted text (titles, paragraphs, footnotes).
+        """
+        from docling.document_converter import DocumentConverter
+
+        idx = int(re.search(r"\d+", image).group()) - 1 if re.search(r"\d+", image) else 0
+        idx = max(0, min(idx, len(self._images) - 1))
+        tmp = self._tmp_dir / f"layout_{idx}.png"
+        self._images[idx].save(tmp)
```
[the remark this answers is not in the document]
Comment (Strongich) [reply trl/experimental/gold/gold_tool_calling_vlm.py:165]: Unreachable; dataset guarantees it.

File: trl/experimental/gold/gold_trainer.py [hunk 4]
```diff
@@ -983,6 +1009,162 @@ def __init__(
...
+
+        # Check for async functions to start an event loop on a daemon thread
+        self._has_async_funcs = any(inspect.iscoroutinefunction(func) for func in self.tools)
+        if self._has_async_funcs:
+            self.async_loop_thread, self.async_loop, self.async_loop_ready_event = start_event_loop_in_daemon(
+                name="GOLDTrainer-AsyncLoop"
+            )
+            # wait until the event loop is running in the daemon thread
+            self.async_loop_ready_event.wait()
+            atexit.register(shutdown_event_loop_in_daemon, self.async_loop_thread, self.async_loop)
```
[the remark this answers is not in the document]
Comment (Strongich) [reply trl/experimental/gold/gold_trainer.py:1142]: FP

File: trl/experimental/gold/gold_trainer.py [hunk 5]
```diff
@@ -983,6 +1009,162 @@ def __init__(
...
+                    "The tokenizer has a legacy `response_schema` set but no `response_template`. The installed "
+                    "transformers supports the new-style `response_template`; consider migrating, as `response_schema` "
+                    "support will eventually be removed. See the Transformers response-parsing docs.",
+                    FutureWarning,
+                )
+        # In multi-turn training, the chat template *must* be prefix-preserving. If the tokenizer's original template
+        # isn't, we replace it with a training-safe, prefix-preserving template. This is (re-)applied *after*
+        # `super().__init__()` below, because `SFTTrainer.__init__` unconditionally sets `self.chat_template` from
+        # `assistant_only_loss` and would otherwise clobber the tool template back to `None`.
+        self.max_tool_calling_iterations = args.max_tool_calling_iterations or sys.maxsize
```
[the remark this answers is not in the document]
Comment (Strongich) [reply trl/experimental/gold/gold_trainer.py]: Addressed in a separate PR https://github.com/huggingface/trl/pull/6423

File: trl/experimental/gold/gold_tool_calling_vlm.py [hunk 6]
```diff
@@ -0,0 +1,339 @@
...
+    def __init__(self):
+        # Built lazily on first use and kept for the life of this (pooled, reused) instance. Rebuilding it per reset
+        # would reload docling's OCR models on every rollout.
+        self._converter = None
+
+    def reset(self, **example):
+        self._images = example.get("images") or []
+        # Per-instance scratch dir: environment instances are pooled and run concurrently, so a shared fixed path
+        # would race between rollouts saving the same image index.
+        self._tmp_dir = Path(tempfile.mkdtemp(prefix="gold_layout_"))
```
[the remark this answers is not in the document]
Comment (Strongich) [reply trl/experimental/gold/gold_tool_calling_vlm.py]: [1b64a13](https://github.com/huggingface/trl/pull/6328/commits/1b64a138cf3d005004e799e219a11572e6de3884)

File: trl/experimental/gold/gold_trainer.py [hunk 7]
```diff
@@ -1621,6 +1819,468 @@ def _generate_on_policy_for_slices(
...
+            ):
+                self.vllm_generation.sync_weights()
+                self._last_vllm_sync_step = self.state.global_step
+
+            # Generate using vLLM with raw token IDs
+            _, completion_ids, _, _ = self.vllm_generation.generate(
+                prompts=prompt_ids,
+                images=images,
+                num_generations=self.num_generations,
+            )
```
[the remark this answers is not in the document]
Comment (Strongich) [reply trl/experimental/gold/gold_trainer.py:1974]: Addressed in a separate PR https://github.com/huggingface/trl/pull/6673
